Automattic/wp-calypso
# Billing History: show receipt item discounts on each item

## Proposed Changes

This PR modifies the billing history receipt view to add a list of each discount applied to its receipt items. This involves reformatting the way that line items are displayed on a receipt in the following ways:

- The line item total was previously the amount paid for that receipt item. Now it is the undiscounted price of the receipt item. That way the discounts which are shown below it will show how they reduced the price from its original total. **This means that the total for each line item is no longer shown anywhere!**
- Taxes were previously shown as subtext below the grand total. Now the tax is shown as a line item above the total. That way the receipt reads like a tally sheet with additions and subtractions to display how the grand total was calculated.

> [!Note]
> The names of each discount for receipts are only shown on receipts which were made after after D126240-code (October 10, 2023) when the data started to be recorded and the amount for each discount is only shown for receipts after D133350-code (January 2, 2024). Receipts before that point should continue to be displayed as they were originally, with the style changes noted above.

> [!Note]
> For introductory offers, this displays the human-readable version of that offer (eg: "First year free" or "Discount for first 3 months"). For offers where the billing term length is different from the product's term length (eg: a 3 month discount for an annual subscription), the discount amount is hidden and replaced with text that reads "Amount paid in transaction: $X" to avoid confusion (if such a discount is -$35, it would be misleading to suggest that the savings is $35 because the customer will be paying $26.25 in 3 months). You can read more about this strategy when we applied it to checkout in https://github.com/Automattic/wp-calypso/pull/87732. See the screenshots below which include such an offer.

Fixes https://github.com/Automattic/wp-calypso/issues/82025

Requires D139433-code

Before             |  After
:-------------------------:|:-------------------------:
<img width="852" alt="Screenshot 2023-11-21 at 10 42 11 AM" src="https://github.com/Automattic/wp-calypso/assets/2036909/61397781-8a0f-402a-a6a7-d3030ba8b87d">  | <img width="846" alt="Screenshot 2023-11-21 at 2 00 55 PM" src="https://github.com/Automattic/wp-calypso/assets/2036909/4fd6e595-cfd4-4bc4-bfb2-ecd9323a6247">
<img width="850" alt="Screenshot 2023-11-21 at 2 06 08 PM" src="https://github.com/Automattic/wp-calypso/assets/2036909/facebedb-596e-4ab8-9d09-85009cc25121"> | <img width="845" alt="Screenshot 2023-11-21 at 2 06 15 PM" src="https://github.com/Automattic/wp-calypso/assets/2036909/0a039843-d3fb-4f3d-a51f-97798d241471">
<img width="958" alt="Screenshot 2024-01-22 at 7 10 15 PM" src="https://github.com/Automattic/wp-calypso/assets/2036909/d71d2f01-e2f9-43f7-ae52-350abd515de4"> | <img width="956" alt="Screenshot 2024-01-22 at 7 10 04 PM" src="https://github.com/Automattic/wp-calypso/assets/2036909/b4a19ff9-a662-43e5-8610-bb60be4c4529">
<img width="657" alt="Screenshot 2024-02-26 at 1 50 22 PM" src="https://github.com/Automattic/wp-calypso/assets/2036909/6e71a21f-538a-43be-9ad1-bbc1c18b3edd"> | <img width="906" alt="Screenshot 2024-03-08 at 7 55 08 PM" src="https://github.com/Automattic/wp-calypso/assets/2036909/df96bf81-1def-47df-8c97-2e643fcc59da">


## Testing Instructions

- Make a purchase with a coupon or find an account which has a purchase that already has a coupon or other discount. It must be a purchase made after D133350-code (January 2, 2024).
- Visit https://wordpress.com/me/purchases/billing
- Click to view the receipt of the purchase with the discount.
- Verify that you see the discount listed under the receipt item and that the math of the prices is correct.
- Look at a purchase with discounts before D133350-code (January 2, 2024) but after D126240-code (October 10, 2023). Verify that the names of the discounts are shown but without amounts (since they might be inaccurate) and that the math of the prices is still correct.
